lib/pry-doc/pry_ext/show_source_with_c_internals/cute_extractor.rb

```diff
@@ -0,0 +1,105 @@
+class CuteExtractor
+  class << self
+    attr_accessor :file_cache
+  end
+  @file_cache = {}
+
+  def extract_code(info)
+    if info.original_symbol.start_with?("#define")
+      extract_macro(info)
+    elsif info.original_symbol =~ /\s*(struct|enum)\s*/
+      extract_struct(info)
+    elsif info.original_symbol.start_with?("}")
+      extract_typedef_struct(info)
+    elsif info.original_symbol =~/^typedef.*;$/
+      extract_typedef_oneliner(info)
+    else
+      extract_function(info)
+    end
+  end
+
+  private
+
+  def extract_struct(info)
+    source_file = source_from_file(info.file)
+    loop do
+      code = source_file[info.line, offset].join
+      return code if balanced?(code)
+      offset += 1
+    end
+  end
+
+  def extract_typedef_struct(info)
+    source_file = source_from_file(info.file)
+    offset = 1
+    loop do
```

Comment:
@kyrylo couldn't come up with a decent extraction here as many of the loops are subtly different, you can try homie